src/Raven.Server/Documents/Handlers/Batches/BatchRequestParser.cs

```diff
@@ -1544,13 +1206,18 @@ private static unsafe void ThrowInvalidCommandType(JsonParserState state, JsonOp
                                                   + ctx.AllocateStringValue(null, state.StringBuffer, state.StringSize));
         }
 
-        private static void ThrowUnexpectedToken(JsonParserToken jsonParserToken, JsonParserState state)
+        public static void ThrowUnexpectedToken(JsonParserToken jsonParserToken, JsonParserState state)
         {
             throw new InvalidOperationException("Expected " + jsonParserToken + " , but got " + state.CurrentTokenType);
         }
 
-        private static async Task RefillParserBuffer(Stream stream, JsonOperationContext.MemoryBuffer buffer, UnmanagedJsonParser parser, CancellationToken token = default)
+
+        public AbstractBatchCommandParsingObserver CommandParsingObserver { get; set; }
+
+        public async Task RefillParserBuffer(Stream stream, JsonOperationContext.MemoryBuffer buffer, UnmanagedJsonParser parser, CancellationToken token = default)
         {
+            CommandParsingObserver?.OnParserBufferRefill(parser);
```

Comment:
Here you aren't passing the `buffer` to the parser, likely causing RavenDB-18408 @shiranshalom

Reply:
Can you elaborate? I don't follow what's a potential issue here.

The idea is that this allows us to copy the already parsed part of JSON command to memory. We're passing just the buffer since the implementation is just `parser.CopyParsedChunk(CommandStream, _commandStartBufferPosition);` and it copies data from the buffer that the parser already has.

https://github.com/ravendb/ravendb/pull/13962/files#diff-a986875160cb5bacd8bc22332d134d3a34b84c88c6845d0558c8fa549e370f99R47

Reply:
This is not related to RavenDB-18408

https://issues.hibernatingrhinos.com/issue/RavenDB-18408#focus=Comments-67-289481.0-0

Reply:
You have `JsonOperationContext.MemoryBuffer buffer` that is passed to the function, but you are then calling to `OnParserBufferRefill` without the buffer, is that okay to do? Won't you miss something because of that?

Reply:
I think it's good. We won't miss anything, unless I don't see something. It works as follow:

- we have `JsonOperationContext.MemoryBuffer buffer` which is filled by data from the request stream
- the parses uses this buffer - we call `parser.SetBuffer(buffer, ...); `
- in general we copy the parsed command once its parsing is completed `CommandParsingObserver?.OnCommandEnd()` 
- but it might happen that we get here - `RefillParserBuffer()` - because we need to refill the buffer in the middle of parsing of a single command. then we need to copy the already parsed part - which is still available in the buffer - `parser.CopyParsedChunk()` does it - and then we can fill / overwrite the buffer by new data from the stream:

https://github.com/ravendb/ravendb/blob/cab987572f8708c6f668c6a4be39ce16bb7d6f27/src/Raven.Server/Documents/Handlers/Batches/BatchRequestParser.cs#L1223-L1226